### DISCUSSION TITLE
Use pipenv with internal nexus pypi mirror

### DISCUSSION THREAD

**Discussion by smutel** [pypa/pipenv#5035] [Q&A] [answered]
Hello,

I tried to find the information between several discussions and issues in this repo but I did'nt find a clear answer so I ask it here.

Is-it possible to use internal pypi repo with pipenv ?
I tried several configurations and it seems that pipenv is always looking to pypi.org.

Thanks.

**Comment by matteius** [comment, marked as answer]
Hi @smutel The first index specified in your Pipfile will be the default.   There are some new documentation updates in master that will be published when there is a new release soon, but basically if you omit PyPI or remove it from your Pipfile and specify your private pypi repo as the first source within the Pipfile, all unspecified packages will use that index by default.  

EDIT:  version 2022.4.8 works for this case because it is more strict about determining if an index is in-fact pypi.

**Reply by smutel** [reply]
Thanks for this info. I just removed the source in my pipfile and it seems to not work as expected. Perhaps I did a mistake somewhere ...

**Here is my /etc/pip.conf:**
```ini
[global]
index = https://xxx/nexus/repository/pypi.org/pypi
index-url = https://xxx/nexus/repository/pypi.org/simple
```
I tested the installation through pip, it is working fine.

**Here is my Pipfile:**
```ini
[dev-packages]
ansible-lint = "==4.3.7"
rich = "<11.0.0"

[packages]
ansible = "~=4.10"
requests = "*"
pynetbox = "*"
dnspython = "*"

[requires]
python_version = "3"
```
This Pipfile worked correctly with default source pointing to pypi.org.

Now I encounter this error message:
```
Pipfile.lock not found, creating...
Locking [dev-packages] dependencies...
Building requirements...
Resolving dependencies...
✘ Locking Failed! 

CRITICAL:pipenv.patched.notpip._internal.resolution.resolvelib.factory:Could not find a version that satisfies the requirement rich<11.0.0 (from versions: none)
[ResolutionFailure]:   File "/usr/local/lib/python3.9/dist-packages/pipenv/resolver.py", line 743, in _main
[ResolutionFailure]:       resolve_packages(pre, clear, verbose, system, write, requirements_dir, packages, dev)
[ResolutionFailure]:   File "/usr/local/lib/python3.9/dist-packages/pipenv/resolver.py", line 704, in resolve_packages
[ResolutionFailure]:       results, resolver = resolve(
[ResolutionFailure]:   File "/usr/local/lib/python3.9/dist-packages/pipenv/resolver.py", line 685, in resolve
[ResolutionFailure]:       return resolve_deps(
[ResolutionFailure]:   File "/usr/local/lib/python3.9/dist-packages/pipenv/utils.py", line 1398, in resolve_deps
[ResolutionFailure]:       results, hashes, markers_lookup, resolver, skipped = actually_resolve_deps(
[ResolutionFailure]:   File "/usr/local/lib/python3.9/dist-packages/pipenv/utils.py", line 1127, in actually_resolve_deps
[ResolutionFailure]:       resolver.resolve()
[ResolutionFailure]:   File "/usr/local/lib/python3.9/dist-packages/pipenv/utils.py", line 905, in resolve
[ResolutionFailure]:       raise ResolutionFailure(message=str(e))
[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  You can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
ERROR: No matching distribution found for rich<11.0.0
```

**Reply by matteius** [reply]
@smutel I think that it will not work this way with the source in  /etc/pip.conf ... at least I am not familiar with configuring it this way.  Instead add the default source to your Pipfile like this:

```
[[source]]
url = "https://xxx/nexus/repository/pypi.org/simple"
verify_ssl = true
name = "internal-pypi"

[dev-packages]
ansible-lint = "==4.3.7"
rich = "<11.0.0"

[packages]
ansible = "~=4.10"
requests = "*"
pynetbox = "*"
dnspython = "*"

[requires]
python_version = "3"
```

**Reply by smutel** [reply]
Even with this configuration, I have this error message:
```
pipenv.vendor.urllib3.exceptions.ConnectTimeoutError: (<pipenv.vendor.urllib3.connection.HTTPSConnection object at 0x7f27eec389d0>, 'Connection to pypi.org timed out. (connect timeout=10)')
```
which indicate that pipenv is looking at pypi.org anyway ...

**Reply by matteius** [reply]
@smutel It is because pypi.org is in your URL for the sources, so it thinks you are in fact trying to reach pypi.org I suspect.

**Reply by matteius** [reply]
@smutel There is a chance that version `2022.4.8` will work better for your case, a chance because it happens I refactored this bit here:  https://github.com/pypa/pipenv/pull/5029/files#diff-62e85156fc3b274046cc03dabfe6e3ceead503f330b98fb3611645216b4b6e1fL747-R747

It looks like the new utility is more strict on how it matches if its truly pypi.org.

**Reply by smutel** [reply]
It works fine with the last version. Thanks.

**Reply by kalebmckale** [reply]
There have been a number of updates including interfacing with the indexes in the `pip.conf` file and adding them to `Pipfile` that now better handles your scenario.
